proxy/src/main/java/com/velocitypowered/proxy/command/CommandGraphInjector.java

```diff
@@ -0,0 +1,142 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package com.velocitypowered.proxy.command;
+
+import com.google.common.base.Preconditions;
+import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.StringReader;
+import com.mojang.brigadier.builder.ArgumentBuilder;
+import com.mojang.brigadier.context.CommandContextBuilder;
+import com.mojang.brigadier.context.StringRange;
+import com.mojang.brigadier.tree.CommandNode;
+import com.mojang.brigadier.tree.LiteralCommandNode;
+import com.mojang.brigadier.tree.RootCommandNode;
+import com.velocitypowered.proxy.command.brigadier.VelocityArgumentCommandNode;
+import java.util.concurrent.locks.Lock;
+import org.checkerframework.checker.lock.qual.GuardedBy;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Copies the nodes of a {@link RootCommandNode} to a possibly non-empty
+ * destination {@link RootCommandNode}, respecting the requirements satisfied
+ * by a given command source.
+ *
+ * @param <S> the type of the source to inject the nodes for
+ */
+public final class CommandGraphInjector<S> {
+
+  private static final StringRange ALIAS_RANGE = StringRange.at(0);
+  private static final StringReader ALIAS_READER = new StringReader("");
+
+  private final @GuardedBy("lock") CommandDispatcher<S> dispatcher;
+  private final Lock lock;
+
+  CommandGraphInjector(final CommandDispatcher<S> dispatcher, final Lock lock) {
+    this.dispatcher = Preconditions.checkNotNull(dispatcher, "dispatcher");
+    this.lock = Preconditions.checkNotNull(lock, "lock");
+  }
+
+  // The term "source" is ambiguous here. We use "origin" when referring to
+  // the root node we are copying nodes from to the destination node.
+
+  /**
+   * Adds the node from the root node of this injector to the given root node,
+   * respecting the requirements satisfied by the given source.
+   *
+   * <p>Prior to adding a literal with the same name as one previously contained
+   * in the destination node, the old node is removed from the destination node.
+   *
+   * @param dest the root node to add the permissible nodes to
+   * @param source the command source to inject the nodes for
+   */
+  public void inject(final RootCommandNode<S> dest, final S source) {
+    lock.lock();
+    try {
+      final RootCommandNode<S> origin = this.dispatcher.getRoot();
```

Comment:
I generally am uneasy with running "alien code" we don't trust in critical sections like this. Can this be modified to clone the `RootCommandNode` in its entirety instead?

Reply:
(Likewise, the `SuggestionsProvider` does something similar. I'm not sure about the best way to handle that.)

Reply:
Now that I think about it, this is why you wanted the precondition in `InvocableCommand` for #515 - because as designed, this will cause a deadlock. This neatly explains why.